### PR TITLE
[candi] refactor kubernetes-api-proxy

### PR DESCRIPTION
**Pull request by RomanenkoDenys**
## Description

Refactor kubernetes-api-proxy.
1. Use discovered apiserver endpoints to set up kubernetes api proxy on nodes.
2. Use kubernetes api proxy static pod instead of system package.
## Why do we need it, and what problem does it solve?

This PR prevents situations, when all apiservers are going down and kubernetes-api-proxy cannot discover live endpoints.
## Changelog entries

```changes
section: candi
type: feature
summary: refactor kubernetes-api-proxy
impact_level: low
```

### CODE REVIEW

File: modules/040-control-plane-manager/images/kubernetes-api-proxy/werf.inc.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+---
+image: {{ $.ModuleName }}/{{ $.ImageName }}
+from: {{ env "BASE_NGINX_ALPINE" }}
+import:
+- image: common/pause
```
Comment (nabokihms) [inline modules/040-control-plane-manager/images/kubernetes-api-proxy/werf.inc.yaml:5]: I assume that pause binary is not required for this image.

File: modules/040-node-manager/hooks/discover_apiserver_endpoints.go [hunk 2]
```diff
@@ -47,45 +68,79 @@ var _ = sdk.RegisterFunc(&go_hook.HookConfig{
...
+	}
+	return addresses, nil
 }
 
 func handleAPIEndpoints(input *go_hook.HookInput) error {
-	snap := input.Snapshots["kube_api_ep"]
-	if len(snap) == 0 {
-		input.LogEntry.Error("kubernetes endpoints not found")
-		return nil
+	endpointsSet := make(map[string]struct{})
```
Comment (nabokihms) [inline modules/040-node-manager/hooks/discover_apiserver_endpoints.go:120]: Use [`set`](https://github.com/deckhouse/deckhouse/blob/8527bd90b654ccc16314d440cf9fda349f2ddd4d/go_lib/set/set.go#L26) here

File: modules/040-node-manager/hooks/discover_apiserver_endpoints.go [hunk 3]
```diff
@@ -47,45 +68,79 @@ var _ = sdk.RegisterFunc(&go_hook.HookConfig{
...
+			if port.Name == "https" {
+				ports = append(ports, port.Port)
 			}
 		}
-	}
 
-	return apiEP, nil
+		for _, addrObj := range s.Addresses {
+			for _, port := range ports {
+				addr := fmt.Sprintf("%s:%d", addrObj.IP, port)
```
Comment (nabokihms) [inline modules/040-node-manager/hooks/discover_apiserver_endpoints.go:111]: https://pkg.go.dev/net#JoinHostPort

**Review by nabokihms — CHANGES_REQUESTED** [review]
Looks great! I have tiny additions to your PR, but anyway, it's good.

File: modules/040-node-manager/hooks/discover_apiserver_endpoints.go (hunk 2 above)
Comment (RomanenkoDenys) [reply modules/040-node-manager/hooks/discover_apiserver_endpoints.go:120]: Thanks !

File: modules/040-node-manager/hooks/discover_apiserver_endpoints.go (hunk 3 above)
Comment (RomanenkoDenys) [reply modules/040-node-manager/hooks/discover_apiserver_endpoints.go:111]: done

File: modules/040-control-plane-manager/images/kubernetes-api-proxy/werf.inc.yaml (hunk 1 above)
Comment (RomanenkoDenys) [reply modules/040-control-plane-manager/images/kubernetes-api-proxy/werf.inc.yaml]: yep, thanks !

**Review by nabokihms — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @nabokihms. I've started the workflow for label 'e2e/run/aws'...

**Pull request merged**